test/tools/ossfuzz/protoToAbiV2.h

```diff
@@ -0,0 +1,368 @@
+	unsigned getNextCounter()
+	{
+		return m_counter++;
+	}
+
+	// m_varCounter is used to derive declared and parameterized variable names.
+	unsigned getNextVarCounter()
+	{
+		return m_varCounter++;
+	}
+
+	// Accepts an unsigned counter and returns a pair of strings.
+	// First string is declared name (s_varNamePrefix<varcounter_value>)
+	// Second string is parameterized name (s_paramPrefix<varcounter_value>)
+	auto newVarNames(unsigned _varCounter)
+	{
+		return std::make_pair(
+			s_varNamePrefix + std::to_string(_varCounter),
+			s_paramNamePrefix + std::to_string(_varCounter)
+		);
+	}
+
+	// String and bytes literals are derived by hashing a monotonically increasing
+	// counter and enclosing the said hash inside double quotes.
+	std::string bytesArrayValueAsString()
+	{
+		return "\"" + toHex(hashUnsignedInt(getNextCounter()), HexPrefix::DontAdd) + "\"";
+	}
+
+	std::string getQualifier(DataType _dataType)
+	{
+		return ((isValueType(_dataType) || m_isStateVar) ? "" : "memory");
+	}
+
+	// Static declarations
+	static std::string structTypeAsString(StructType const& _x);
+	static std::string intValueAsString(unsigned _width, unsigned _counter);
+	static std::string uintValueAsString(unsigned _width, unsigned _counter);
+	static std::string integerValueAsString(bool _sign, unsigned _width, unsigned _counter);
+	static std::string addressValueAsString(unsigned _counter);
+	static std::string fixedByteValueAsString(unsigned _width, unsigned _counter);
+	static std::vector<std::pair<bool, unsigned>> arrayDimensionsAsPairVector(ArrayType const& _x);
+	static std::string arrayDimInfoAsString(ArrayDimensionInfo const& _x);
+	static void arrayDimensionsAsStringVector(
+		ArrayType const& _x,
+		std::vector<std::string>&
+	);
+	static std::string bytesArrayTypeAsString(DynamicByteArrayType const& _x);
+	static std::string arrayTypeAsString(std::string const&, ArrayType const&);
+	static std::string delimiterToString(Delimiter _delimiter);
+
+	// Static function definitions
+	static bool isValueType(DataType _dataType)
+	{
+		return _dataType == DataType::VALUE;
+	}
+
+	static unsigned getIntWidth(IntegerType const& _x)
+	{
+		return 8 * ((_x.width() % 32) + 1);
+	}
+
+	static bool isIntSigned(IntegerType const& _x)
+	{
+		return _x.is_signed();
+	}
+
+	static std::string getIntTypeAsString(IntegerType const& _x)
+	{
+		return ((isIntSigned(_x) ? "int" : "uint") + std::to_string(getIntWidth(_x)));
+	}
+
+	static unsigned getFixedByteWidth(FixedByteType const& _x)
+	{
+		return (_x.width() % 32) + 1;
+	}
+
+	static std::string getFixedByteTypeAsString(FixedByteType const& _x)
+	{
+		return "bytes" + std::to_string(getFixedByteWidth(_x));
+	}
+
+	static std::string getAddressTypeAsString(AddressType const& _x)
+	{
+		return (_x.payable() ? "address payable": "address");
+	}
+
+	// Convert _counter to string and return its keccak256 hash.
+	static u256 hashUnsignedInt(unsigned _counter)
+	{
+		return keccak256(
+			boost::algorithm::to_lower_copy(
```

Comment:
`to_string` produces a decimal number - why do you call `to_lower_copy`?

You could also use `keccak256(h256(_counter))`.

Reply:
I borrowed the code snippet from `libdevcore` code that was hashing strings and used it as-is.